src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueue.java

```diff
@@ -547,23 +548,43 @@ public WriteLock appendLock() {
     @NotNull
     @Override
     public ExcerptTailer createTailer(String id) {
-        throwExceptionIfClosed();
+        verifyTailerPreconditions(id);
 
         LongValue index = id == null
                 ? null
                 : indexForId(id);
-        final StoreTailer storeTailer = new StoreTailer(this, pool, index);
+        LongValue indexVersion = id == null
+                ? null
+                : indexVersionForId(id);
+
+        final StoreTailer storeTailer = new StoreTailer(this, pool, index, indexVersion);
         directoryListing.refresh(true);
         storeTailer.singleThreadedCheckReset();
         return storeTailer;
     }
 
+    private void verifyTailerPreconditions(String id) {
+        // Preconditions for all tailer types
+        throwExceptionIfClosed();
+
+        // Named tailer preconditions
+        if (id == null) return;
+        if (appendLock.locked() && id.startsWith(REPLICATED_NAMED_TAILER_PREFIX)) {
+            throw new NamedTailerNotAvailableException(id, NamedTailerNotAvailableException.Reason.NOT_AVAILABLE_ON_SINK);
+        }
+    }
+
     @Override
     @NotNull
     public LongValue indexForId(@NotNull String id) {
         return this.metaStore.doWithExclusiveLock((ts) -> ts.acquireValueFor("index." + id, 0L));
     }
 
+    @NotNull
+    public LongValue indexVersionForId(@NotNull String id) {
+        return this.metaStore.doWithExclusiveLock((ts) -> ts.acquireValueFor("index." + id + ".version", -1L));
+    }
```

Comment:
We don't really need to have a version for non-replicated named tailers. I wonder if it's worth having an abstraction e.g. IndexUpdater and it bumps the index and the version for a replicated tailer, but only writes the index for a non-replicated one. It might be easier to just have them both write the version? up to you.

Reply:
Hi @nicktindall thanks for this - I think you're right and that this would be an improvement. I have added the `IndexUpdater` and `IndexUpdaterFactory` classes that implement this. This does seem cleaner and minimises the impact to existing named tailers. Moreover we avoid the cost of the version increment in non replicated tailers.

Reply:
In the future I would like to come back and resolve all of the null checking in StoreTailer - for example a null index was historically passed into StoreTailer to indicate an unnamed tailer. The new additions still conform to this design of passing null around. I would like to come back and refactor this after we have released this.